File: src/mastra/tools/sql-generation-tool.ts
import { createTool } from '@mastra/core/tools';
import { z } from 'zod';
import mysql from 'mysql2/promise';

export const sqlGenerationTool = createTool({
  id: 'sql-generation',
  inputSchema: z.object({
    question: z.string().describe('Natural language question about colleges and admissions'),
    generateOnly: z
      .boolean()
      .optional()
      .describe('If true, only generate SQL without executing. If false, generate and execute.'),
  }),
  description:
    'Generates SQL SELECT queries from natural language questions. First introspects database schema, then generates appropriate SQL.',
  execute: async inputData => {
    const { question, generateOnly = false } = inputData;
    const dbUrl = process.env.DATABASE_URL;

    if (!dbUrl) {
      throw new Error('DATABASE_URL environment variable not set');
    }

    let connection: mysql.Connection | null = null;

    try {
      console.log(`📝 Processing question: "${question}"\n`);

      connection = await mysql.createConnection({
        uri: dbUrl,
        connectTimeout: 30000,
      });

      // Get database name
      const [dbResult] = await connection.execute('SELECT DATABASE() as db_name');
      const currentDb = (dbResult as any[])[0]?.db_name;
      console.log(`📊 Database: ${currentDb}\n`);

      // Get table information
      const [tables] = await connection.execute(
        `SELECT TABLE_NAME FROM information_schema.TABLES 
         WHERE TABLE_SCHEMA = ? ORDER BY TABLE_NAME`,
        [currentDb],
      );

      const tableNames = (tables as any[]).map(t => t.TABLE_NAME);
      console.log(`📋 Available tables: ${tableNames.join(', ')}\n`);

      // Get columns information
      const [columns] = await connection.execute(
        `SELECT TABLE_NAME, COLUMN_NAME, DATA_TYPE, COLUMN_KEY, IS_NULLABLE
         FROM information_schema.COLUMNS 
         WHERE TABLE_SCHEMA = ? 
         ORDER BY TABLE_NAME, ORDINAL_POSITION`,
        [currentDb],
      );

      // Build schema context
      const schemaContext = buildSchemaText(columns as any[]);

      return {
        success: true,
        question,
        database: currentDb,
        tables: tableNames,
        schemaContext,
        nextInstruction:
          'Now use sql-execution tool to execute the query. The LLM should generate a SELECT query based on the schema context.',
      };
    } catch (error) {
      throw new Error(`Failed to generate SQL: ${error instanceof Error ? error.message : String(error)}`);
    } finally {
      if (connection) {
        await connection.end();
      }
    }
  },
});

function buildSchemaText(columns: any[]): string {
  const tableColumns = new Map<string, any[]>();

  columns.forEach((col: any) => {
    if (!tableColumns.has(col.TABLE_NAME)) {
      tableColumns.set(col.TABLE_NAME, []);
    }
    tableColumns.get(col.TABLE_NAME)?.push(col);
  });

  let text = 'DATABASE SCHEMA:\n\n';

  for (const [tableName, cols] of tableColumns) {
    text += `TABLE: ${tableName}\n`;
    cols.forEach((col: any) => {
      const pk = col.COLUMN_KEY === 'PRI' ? ' [PRIMARY KEY]' : '';
      const nullable = col.IS_NULLABLE === 'YES' ? '' : ' [NOT NULL]';
      text += `  - ${col.COLUMN_NAME}: ${col.DATA_TYPE}${pk}${nullable}\n`;
    });
    text += '\n';
  }

  return text;
}

File: src/tools/sql_generation_tool.py
"""SQL generation tool - converts natural language to SQL"""

import os
from typing import Optional

import mysql.connector
from mysql.connector import MySQLConnection


async def create_database_connection(connection_string: str) -> MySQLConnection:
    """Create a MySQL connection"""
    return mysql.connector.connect(
        uri=connection_string,
        connection_timeout=30,
    )


async def execute_query(connection: MySQLConnection, query: str) -> list:
    """Execute a query and return results"""
    cursor = connection.cursor(dictionary=True)
    try:
        cursor.execute(query)
        return cursor.fetchall()
    finally:
        cursor.close()


def build_schema_text(columns: list) -> str:
    """Build a text representation of the database schema"""
    table_columns = {}

    for col in columns:
        table_name = col.get("table_name") or col.get("TABLE_NAME")
        if table_name not in table_columns:
            table_columns[table_name] = []
        table_columns[table_name].append(col)

    text = "DATABASE SCHEMA:\n\n"

    for table_name in sorted(table_columns.keys()):
        cols = table_columns[table_name]
        text += f"TABLE: {table_name}\n"

        for col in cols:
            col_name = col.get("column_name") or col.get("COLUMN_NAME")
            data_type = col.get("data_type") or col.get("DATA_TYPE")
            column_key = col.get("column_key") or col.get("COLUMN_KEY")
            is_nullable = col.get("is_nullable") or col.get("IS_NULLABLE")

            pk = " [PRIMARY KEY]" if column_key == "PRI" else ""
            nullable = "" if is_nullable == "YES" else " [NOT NULL]"

            text += f"  - {col_name}: {data_type}{pk}{nullable}\n"

        text += "\n"

    return text


async def sql_generation_tool(
    question: str,
    generate_only: bool = False,
    database_url: Optional[str] = None,
) -> dict:
    """
    Generate SQL SELECT queries from natural language questions.
    First introspects database schema, then generates appropriate SQL.

    Args:
        question: Natural language question about colleges and admissions
        generate_only: If True, only generate SQL without executing
        database_url: Database connection URL (uses DATABASE_URL env var if not provided)

    Returns:
        dict: Generated SQL and schema information
    """
    db_url = database_url or os.getenv("DATABASE_URL")

    if not db_url:
        raise ValueError("DATABASE_URL environment variable not set")

    connection: Optional[MySQLConnection] = None

    try:
        print(f'📝 Processing question: "{question}"\n')

        connection = await create_database_connection(db_url)

        # Get database name
        db_result = await execute_query(connection, "SELECT DATABASE() as db_name")
        current_db = db_result[0]["db_name"] if db_result else None
        print(f"📊 Database: {current_db}\n")

        # Get table information
        tables_result = await execute_query(
            connection,
            """
            SELECT TABLE_NAME FROM information_schema.TABLES
            WHERE TABLE_SCHEMA = DATABASE()
            ORDER BY TABLE_NAME
            """,
        )

        table_names = [t["TABLE_NAME"] for t in tables_result]
        print(f"📋 Available tables: {', '.join(table_names)}\n")

        # Get columns information
        columns_result = await execute_query(
            connection,
            """
            SELECT TABLE_NAME, COLUMN_NAME, DATA_TYPE, COLUMN_KEY, IS_NULLABLE
            FROM information_schema.COLUMNS
            WHERE TABLE_SCHEMA = DATABASE()
            ORDER BY TABLE_NAME, ORDINAL_POSITION
            """,
        )

        # Build schema context
        schema_context = build_schema_text(columns_result)

        return {
            "success": True,
            "question": question,
            "database": current_db,
            "tables": table_names,
            "schema_context": schema_context,
            "generate_only": generate_only,
            "nextInstruction": "Now use sql-execution tool to execute the query. The LLM should generate a SELECT query based on the schema context.",
        }

    except Exception as error:
        raise RuntimeError(f"Failed to generate SQL: {str(error)}") from error
    finally:
        if connection:
            connection.close()
